Entity/Components/StatusComponent.gd

```diff
@@ -3,34 +3,49 @@ class_name StatusComponent
 ## Applies status to the entity and holds a list of current status.
 
 
-var current_status: Array[StatusBase]
+var current_on_turn_start_status: Array[StatusBase] = []
+var current_on_apply_status: Array[StatusBase] = []
```

Comment:
What's the purpose of splitting these arrays? Could we just assume that any status could do both of these and call the functions for all of them?

Reply:
The idea behind splitting is that at the start of the turn, you can do one loop on the turn_start array, apply the effect and reduce by one the duration, and one more loop on the apply_status to just decrease the duration
This allows to not have to check if each status is on_apply. But this was before I even made the `on_apply` boolean for each status. In the current structure, we could totally have a single status_array and check for each status if they have an effect at_turn_start (might need to add a boolean for that though).

The current structure has one flaw, it doesn't allow effects to both have an effect on apply and at turn start, but I didn't find an effect that did such thing (i'm not sure if poison counts ? because technically it's the effect that adds poison stacks, not the status itself that adds poison stacks on apply, and thus it would just be a turn_start status ?)

Reply:
I think it'd be good to combine them into a single array like you mentioned.

I don't think there's a status that does both currently. `turn_start` would be stuff that gets applied each turn, like poison. `on_apply` would be stuff that gets applied once, like a strength buff that lasts for the entire encounter.

I'm not sure I see the reasoning for using a bool. Couldn't we always call `turn_start` or `on_apply` for every status, but the function is left empty if that status doesn't do anything for that specific event?

Reply:
The logic behind adding a boolean for `on_turn_start` was to not assume that every non `on_apply` status was necessarily a `on_turn_start` status.

Reply:
Right, but is that really necessary? We could just assume that any status can do any of those functions if it wants, and then implement it in that specific status.

The reason I'm worried is because we may have many more conditions for applying status. For example:
- on_turn_start
- on_application
- on_take_damage
- on_card_drawn
- on_heal

With the current implementation, wouldn't we have to have a boolean for each of these different conditions?

Reply:
Not necessarily, if they handle in the same way. The reason we need one for `on_apply` and one for `on_turn_start` is that they should be dealt with differently.
`on_turn_start` should have their duration decreased and their effect applied at the start of each turn;
whereas `on_apply` should have their duration decreased at the start of each turn, but they are not applied on each turn, just at the beginning.
`on_take_damage` would probably be an `on_apply` effect
same for `on_card_draw` and `on_heal` actually. Those wouldn't even be inside the status actually. Because let's imagine you want to apply a status when you are healed, this is probably the effect of a card, so it's the role of the card to do it.
If it's more something like getting more heal when you receive a heal, then that's perfectly handled by the stat system already, and same thing for `on_card_drawn` and `on_take_damage`

Reply:
I'm not sure how those would fit into those categories. Maybe we are thinking about things differently. Here's some examples of what I'm thinking those events would do:
- "When healed, draw 2 cards" - calls `CardManager.card_container.draw_cards(2)` in the `on_heal` function of the status
- "When a card is drawn, deal 1 damage to all enemies" - calls `deal_damage` on all enemies from the `on_card_drawn` function of the status
- "Take damage at the start of your turn" (aka poison) - calls `deal_damage` on the owner of the status in `on_turn_start`
- "When taking damage, deal 1 damage back to the attacker" - calls `deal_damage` on the attacker in `on_take_damage` of the status
- "Gain 2 strength" - calls `_modify_stats_with_status` in `on_apply`

Unless we want each status to handle durations differently, then every status could just decrease its duration in `on_turn_start`.

Was that what you had in mind? Or were you thinking about things differently?

Reply:
Why would we bind signals to the `on_apply` ? And yes, I think binding in the child Status would make more sense indeed

The `modifier_name` is set in the `_init()` and not the in the `init_status()` (I think name should probably be changed for clarity btw). I think I prefer having the boolean for now, as I don't really understand the solution you are giving (maybe the lack of sleep is not helping)

Reply:
I guess we could bind the signals in `_init` or `init_status` too, as long as we just bind them once at the start.

I see, now that I look at it `_init` and `init_status` are pretty much the same thing right? Both get called once when the status is first initialized. Maybe we could remove `_init` all together for clarity?

As for my solution I was talking about, here's what I mean. We would change StatusBase to this:
```GDScript
# StatusBase.gd
func on_apply() -> void:
   pass
```
We would also remove the `is_stat_modification` and `is_on_apply` bools. Then in children of StatusBase we override on_apply and do something like:
```GDScript
# Child of StatusBase.gd
func init_status(in_caster: Entity, in_target: Entity) -> void:
   status_caster = in_caster
   status_target = in_target
   targeted_modifier_dict = in_target.get_stat_component().get_stats().offense_modifier_dict
   modifier_name = StatDictBase.POSSIBLE_MODIFIER_NAMES.damage

func on_apply() -> void:
   _modify_stats_with_status()
```

But like I said, the bool version works as well, just up to preference

Reply:
The difference between `_init` and `init_status` is that:
- `_init` gets called automatically when you make a new instance of the class (and `init_status` doesn't)
- `_init` is being used for things that we know will always be true (such as the effect of a given status)
- `init_status` is being used to set caster and target (which is dependent on the situation). I would actually rename `init_status` to `set_caster_target` or `set_status_runtime_parameters`